File: day05.nim
import std/enumerate
import std/sugar
import std/strutils
import std/re
import std/algorithm
import std/sequtils

proc parseFile(path: string): (seq[string], seq[string]) =
  let f = open(path)
  let contents = f.readAll().split("\n\n").map(text => text.splitlines())
  return (contents[0], contents[1])


proc stacksFromLayout(crateLayout: seq[string]): seq[seq[char]] =
  var stacks = newSeqWith(10, newSeq[char]())
  let lines = reversed(crateLayout)
  let crates = lines[1..^1]

  for line in crates:
    for i, x in enumerate(1, countup(1, line.len-1, 4)):
      if line[x] != ' ':
          stacks[i].add(line[x])
  return stacks


proc stringFromStackTops(stacks: var seq[seq[char]], instructions: seq[string], part: int): string =
  var ops: seq[int]
  var letters: string
  var amount, source, dest, stackIndex, sourceSize: int

  # updating stack by instructions
  for line in instructions:
    if line.len > 0:
      ops = findAll(line, re"\d+").map(parseInt)
      amount = ops[0]
      source = ops[1]
      dest   = ops[2]

      if part == 2:
        sourceSize = stacks[source].len
        for _ in 1..amount:
          stacks[0].add(stacks[source].pop())
        stackIndex = 0
      else:
        stackIndex = source

      for _ in 1..amount:
        stacks[dest].add(stacks[stackIndex].pop())

  # concatenating letters from top of the stacks
  for i in 1..<stacks.len:
    letters &= stacks[i][^1]
  return letters


let (crateLayout, instructions) = parseFile("05.txt")
var p1stacks = stacksFromLayout(crateLayout)
var p2stacks = deepCopy(p1stacks)

# part 1
echo stringFromStackTops(p1stacks, instructions, 1)
# part 2
echo stringFromStackTops(p2stacks, instructions, 2)

File: day05.rs
fn stacks_from_layout(crate_layout: &str) -> Vec<Vec<char>> {
    let mut stacks = (0..10).map(|_| vec![]).collect::<Vec<Vec<char>>>();
    let crates = crate_layout.split("\n").collect::<Vec<&str>>();
    crates.iter()
          .rev()
          .skip(1)
          .for_each(|&line| line.chars()
                                .skip(1)
                                .step_by(4)
                                .enumerate()
                                .filter(|&(_, c)| c != ' ')
                                .for_each(|(i, c)| stacks[i+1].push(c)));
    stacks
}


fn string_from_stacktops(crates: &str, instructions: &str, part: u8) -> String {
    let mut stacks = stacks_from_layout(crates);
    let lines = instructions.split("\n");
    let mut s = String::new();
    let mut stack_index;
    let mut v;

    for line in lines {
        if line.len() == 0 { break }

        if let &[amount, source, dest] =
            line.split(" ")
                .skip(1)
                .step_by(2)
                .map(|x| x.parse::<usize>().unwrap())
                .collect::<Vec<_>>()
                .as_slice()
        {
            if part == 2 {
                for _ in 0..amount {
                    v = stacks[source].pop().unwrap();
                    stacks[0].push(v);
                }
                stack_index = 0;
            }
            else {
                stack_index = source;
            }

            for _ in 0..amount {
                v = stacks[stack_index].pop().unwrap();
                stacks[dest].push(v);
            }
        }
    }


    for i in 1..10 {
        if let Some(&x) = &stacks[i].last(){
            s.push(x);
        }
    }
    s
}


fn main() {
    let text = std::fs::read_to_string("05.txt").expect("404");
    let (crates, instructions) = text.split_once("\n\n").unwrap();

    println!("{:?}", string_from_stacktops(crates, instructions, 1));
    println!("{:?}", string_from_stacktops(crates, instructions, 2));
}
